File: simulators/java/SensorSimulator.java
import java.io.OutputStream;
import java.net.HttpURLConnection;
import java.net.URL;
import java.nio.charset.StandardCharsets;
import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.util.Random;

/**
 * Sensor Simulator - Simulates IoT sensor behavior
 * 
 * This program simulates sensor behavior by sending fictitious data to the IoT API:
 * - Sends data every ~30 seconds with ±10 seconds random variation
 * - Sends temperature, humidity, and/or pressure (1, 2, or 3 fields randomly)
 * - Sometimes skips sending data (simulating packet loss)
 */
public class SensorSimulator {
    
    private static final String API_URL = "http://localhost:8080/api/sensors";
    private static final String SENSOR_NAME = "Sensor-Simulator-Java";
    private static final String LOCATION = "Test-Lab";
    
    private static final int BASE_INTERVAL = 30; // seconds
    private static final int TIME_VARIATION = 10; // ±10 seconds
    
    private static final Random random = new Random();
    
    public static void main(String[] args) {
        System.out.println("=== IoT Sensor Simulator (Java) ===");
        System.out.println("API URL: " + API_URL);
        System.out.println("Sensor: " + SENSOR_NAME);
        System.out.println("Location: " + LOCATION);
        System.out.println("Interval: ~" + BASE_INTERVAL + "s (±" + TIME_VARIATION + "s)");
        System.out.println("Starting simulation...\n");
        
        int cycleCount = 0;
        
        while (true) {
            try {
                cycleCount++;
                log("Cycle " + cycleCount + " starting...");
                
                // Calculate random wait time (30 ± 10 seconds)
                int waitTime = BASE_INTERVAL + random.nextInt(2 * TIME_VARIATION + 1) - TIME_VARIATION;
                log("Waiting " + waitTime + " seconds...");
                Thread.sleep(waitTime * 1000L);
                
                // Simulate packet loss (10% chance)
                if (random.nextInt(10) == 0) {
                    log("PACKET LOSS - Skipping transmission");
                    System.out.println();
                    continue;
                }
                
                // Generate sensor data
                SensorData data = generateSensorData();
                
                // Send data to API
                sendData(data);
                System.out.println();
                
            } catch (InterruptedException e) {
                log("Simulation interrupted");
                break;
            } catch (Exception e) {
                log("ERROR: " + e.getMessage());
                System.out.println();
            }
        }
    }
    
    /**
     * Generate random sensor data with 1, 2, or 3 fields
     */
    private static SensorData generateSensorData() {
        // Decide how many fields to send (1, 2, or 3)
        int fieldCount = random.nextInt(3) + 1;
        
        SensorData data = new SensorData();
        data.name = SENSOR_NAME;
        data.location = LOCATION;
        
        if (fieldCount == 1) {
            // Send only one field
            int field = random.nextInt(3);
            switch (field) {
                case 0:
                    data.temperature = generateTemperature();
                    log("Generated data: Temperature=" + String.format("%.2f", data.temperature) + "°C");
                    break;
                case 1:
                    data.humidity = generateHumidity();
                    log("Generated data: Humidity=" + String.format("%.2f", data.humidity) + "%");
                    break;
                case 2:
                    data.pressure = generatePressure();
                    log("Generated data: Pressure=" + String.format("%.2f", data.pressure) + " hPa");
                    break;
            }
        } else if (fieldCount == 2) {
            // Send two fields
            int excluded = random.nextInt(3);
            if (excluded != 0) {
                data.temperature = generateTemperature();
            }
            if (excluded != 1) {
                data.humidity = generateHumidity();
            }
            if (excluded != 2) {
                data.pressure = generatePressure();
            }
            log("Generated data: " + formatData(data));
        } else {
            // Send all three fields
            data.temperature = generateTemperature();
            data.humidity = generateHumidity();
            data.pressure = generatePressure();
            log("Generated data: " + formatData(data));
        }
        
        return data;
    }
    
    /**
     * Generate random temperature (15-30°C)
     */
    private static double generateTemperature() {
        return 15.0 + random.nextDouble() * 15.0;
    }
    
    /**
     * Generate random humidity (30-80%)
     */
    private static double generateHumidity() {
        return 30.0 + random.nextDouble() * 50.0;
    }
    
    /**
     * Generate random pressure (980-1040 hPa)
     */
    private static double generatePressure() {
        return 980.0 + random.nextDouble() * 60.0;
    }
    
    /**
     * Format sensor data for display
     */
    private static String formatData(SensorData data) {
        StringBuilder sb = new StringBuilder();
        if (data.temperature != null) {
            sb.append("Temperature=").append(String.format("%.2f", data.temperature)).append("°C");
        }
        if (data.humidity != null) {
            if (sb.length() > 0) sb.append(", ");
            sb.append("Humidity=").append(String.format("%.2f", data.humidity)).append("%");
        }
        if (data.pressure != null) {
            if (sb.length() > 0) sb.append(", ");
            sb.append("Pressure=").append(String.format("%.2f", data.pressure)).append(" hPa");
        }
        return sb.toString();
    }
    
    /**
     * Send sensor data to API
     */
    private static void sendData(SensorData data) throws Exception {
        // Build JSON payload
        StringBuilder json = new StringBuilder("{");
        json.append("\"name\":\"").append(data.name).append("\",");
        json.append("\"location\":\"").append(data.location).append("\"");
        
        if (data.temperature != null) {
            json.append(",\"temperature\":").append(data.temperature);
        }
        if (data.humidity != null) {
            json.append(",\"humidity\":").append(data.humidity);
        }
        if (data.pressure != null) {
            json.append(",\"pressure\":").append(data.pressure);
        }
        
        json.append("}");
        
        // Send HTTP POST request
        URL url = new URL(API_URL);
        HttpURLConnection conn = (HttpURLConnection) url.openConnection();
        conn.setRequestMethod("POST");
        conn.setRequestProperty("Content-Type", "application/json");
        conn.setDoOutput(true);
        
        try (OutputStream os = conn.getOutputStream()) {
            byte[] input = json.toString().getBytes(StandardCharsets.UTF_8);
            os.write(input, 0, input.length);
        }
        
        int responseCode = conn.getResponseCode();
        if (responseCode == 201) {
            log("SUCCESS - Data sent to API");
        } else {
            log("WARNING - API returned code: " + responseCode);
        }
        
        conn.disconnect();
    }
    
    /**
     * Log message with timestamp
     */
    private static void log(String message) {
        String timestamp = LocalDateTime.now().format(DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss"));
        System.out.println("[" + timestamp + "] " + message);
    }
    
    /**
     * Data class to hold sensor readings
     */
    static class SensorData {
        String name;
        String location;
        Double temperature;
        Double humidity;
        Double pressure;
    }
}

File: simulators/c/sensor_simulator.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <unistd.h>
#include <curl/curl.h>

/**
 * Sensor Simulator in C - Simulates IoT sensor behavior
 * 
 * This program simulates sensor behavior by sending fictitious data to the IoT API:
 * - Sends data every ~30 seconds with ±10 seconds random variation
 * - Sends temperature, humidity, and/or pressure (1, 2, or 3 fields randomly)
 * - Sometimes skips sending data (simulating packet loss)
 */

#define API_URL "http://localhost:8080/api/sensors"
#define SENSOR_NAME "Sensor-Simulator-C"
#define LOCATION "Test-Lab"
#define BASE_INTERVAL 30
#define TIME_VARIATION 10

typedef struct {
    char name[100];
    char location[100];
    double temperature;
    double humidity;
    double pressure;
    int has_temperature;
    int has_humidity;
    int has_pressure;
} SensorData;

/**
 * Get current timestamp string
 */
void get_timestamp(char *buffer, size_t size) {
    time_t now = time(NULL);
    struct tm *t = localtime(&now);
    strftime(buffer, size, "%Y-%m-%d %H:%M:%S", t);
}

/**
 * Log message with timestamp
 */
void log_message(const char *message) {
    char timestamp[64];
    get_timestamp(timestamp, sizeof(timestamp));
    printf("[%s] %s\n", timestamp, message);
}

/**
 * Generate random double in range [min, max]
 */
double random_double(double min, double max) {
    return min + ((double)rand() / RAND_MAX) * (max - min);
}

/**
 * Generate random temperature (15-30°C)
 */
double generate_temperature() {
    return random_double(15.0, 30.0);
}

/**
 * Generate random humidity (30-80%)
 */
double generate_humidity() {
    return random_double(30.0, 80.0);
}

/**
 * Generate random pressure (980-1040 hPa)
 */
double generate_pressure() {
    return random_double(980.0, 1040.0);
}

/**
 * Generate sensor data with random fields
 */
void generate_sensor_data(SensorData *data) {
    strcpy(data->name, SENSOR_NAME);
    strcpy(data->location, LOCATION);
    
    data->has_temperature = 0;
    data->has_humidity = 0;
    data->has_pressure = 0;
    
    // Decide how many fields to send (1, 2, or 3)
    int field_count = (rand() % 3) + 1;
    
    if (field_count == 1) {
        // Send only one field
        int field = rand() % 3;
        switch (field) {
            case 0:
                data->temperature = generate_temperature();
                data->has_temperature = 1;
                printf("[%s] Generated data: Temperature=%.2f°C\n", "", data->temperature);
                break;
            case 1:
                data->humidity = generate_humidity();
                data->has_humidity = 1;
                printf("[%s] Generated data: Humidity=%.2f%%\n", "", data->humidity);
                break;
            case 2:
                data->pressure = generate_pressure();
                data->has_pressure = 1;
                printf("[%s] Generated data: Pressure=%.2f hPa\n", "", data->pressure);
                break;
        }
    } else if (field_count == 2) {
        // Send two fields
        int excluded = rand() % 3;
        if (excluded != 0) {
            data->temperature = generate_temperature();
            data->has_temperature = 1;
        }
        if (excluded != 1) {
            data->humidity = generate_humidity();
            data->has_humidity = 1;
        }
        if (excluded != 2) {
            data->pressure = generate_pressure();
            data->has_pressure = 1;
        }
    } else {
        // Send all three fields
        data->temperature = generate_temperature();
        data->has_temperature = 1;
        data->humidity = generate_humidity();
        data->has_humidity = 1;
        data->pressure = generate_pressure();
        data->has_pressure = 1;
    }
    
    // Log generated data
    if (field_count > 1) {
        char timestamp[64];
        get_timestamp(timestamp, sizeof(timestamp));
        printf("[%s] Generated data: ", timestamp);
        int first = 1;
        if (data->has_temperature) {
            printf("Temperature=%.2f°C", data->temperature);
            first = 0;
        }
        if (data->has_humidity) {
            if (!first) printf(", ");
            printf("Humidity=%.2f%%", data->humidity);
            first = 0;
        }
        if (data->has_pressure) {
            if (!first) printf(", ");
            printf("Pressure=%.2f hPa", data->pressure);
        }
        printf("\n");
    }
}

/**
 * Build JSON payload from sensor data
 */
void build_json(const SensorData *data, char *json, size_t size) {
    char temp[512];
    snprintf(json, size, "{\"name\":\"%s\",\"location\":\"%s\"", data->name, data->location);
    
    if (data->has_temperature) {
        snprintf(temp, sizeof(temp), ",\"temperature\":%.2f", data->temperature);
        strncat(json, temp, size - strlen(json) - 1);
    }
    if (data->has_humidity) {
        snprintf(temp, sizeof(temp), ",\"humidity\":%.2f", data->humidity);
        strncat(json, temp, size - strlen(json) - 1);
    }
    if (data->has_pressure) {
        snprintf(temp, sizeof(temp), ",\"pressure\":%.2f", data->pressure);
        strncat(json, temp, size - strlen(json) - 1);
    }
    
    strncat(json, "}", size - strlen(json) - 1);
}

/**
 * Send sensor data to API using cURL
 */
int send_data(const SensorData *data) {
    CURL *curl;
    CURLcode res;
    char json[1024] = {0};
    long response_code = 0;
    
    build_json(data, json, sizeof(json));
    
    curl = curl_easy_init();
    if (!curl) {
        log_message("ERROR: Failed to initialize cURL");
        return 0;
    }
    
    struct curl_slist *headers = NULL;
    headers = curl_slist_append(headers, "Content-Type: application/json");
    
    curl_easy_setopt(curl, CURLOPT_URL, API_URL);
    curl_easy_setopt(curl, CURLOPT_POSTFIELDS, json);
    curl_easy_setopt(curl, CURLOPT_HTTPHEADER, headers);
    curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, NULL); // Discard response body
    
    res = curl_easy_perform(curl);
    
    if (res != CURLE_OK) {
        char msg[256];
        snprintf(msg, sizeof(msg), "ERROR: cURL failed: %s", curl_easy_strerror(res));
        log_message(msg);
        curl_easy_cleanup(curl);
        curl_slist_free_all(headers);
        return 0;
    }
    
    curl_easy_getinfo(curl, CURLINFO_RESPONSE_CODE, &response_code);
    
    if (response_code == 201) {
        log_message("SUCCESS - Data sent to API");
    } else {
        char msg[256];
        snprintf(msg, sizeof(msg), "WARNING - API returned code: %ld", response_code);
        log_message(msg);
    }
    
    curl_easy_cleanup(curl);
    curl_slist_free_all(headers);
    
    return (response_code == 201);
}

int main() {
    printf("=== IoT Sensor Simulator (C) ===\n");
    printf("API URL: %s\n", API_URL);
    printf("Sensor: %s\n", SENSOR_NAME);
    printf("Location: %s\n", LOCATION);
    printf("Interval: ~%ds (±%ds)\n", BASE_INTERVAL, TIME_VARIATION);
    printf("Starting simulation...\n\n");
    
    // Initialize random seed
    srand(time(NULL));
    
    // Initialize cURL globally
    curl_global_init(CURL_GLOBAL_ALL);
    
    int cycle_count = 0;
    
    while (1) {
        cycle_count++;
        
        char msg[256];
        snprintf(msg, sizeof(msg), "Cycle %d starting...", cycle_count);
        log_message(msg);
        
        // Calculate random wait time (30 ± 10 seconds)
        int wait_time = BASE_INTERVAL + (rand() % (2 * TIME_VARIATION + 1)) - TIME_VARIATION;
        snprintf(msg, sizeof(msg), "Waiting %d seconds...", wait_time);
        log_message(msg);
        sleep(wait_time);
        
        // Simulate packet loss (10% chance)
        if (rand() % 10 == 0) {
            log_message("PACKET LOSS - Skipping transmission");
            printf("\n");
            continue;
        }
        
        // Generate and send sensor data
        SensorData data;
        generate_sensor_data(&data);
        send_data(&data);
        printf("\n");
    }
    
    curl_global_cleanup();
    return 0;
}
